topology/graph/client.go

```diff
@@ -169,8 +173,12 @@ func (c *AsyncClient) AddListener(l EventListener) {
 
 func (c *AsyncClient) Disconnect() {
 	c.running.Store(false)
-	c.quit <- true
+	if c.connected.Load() == true {
+		c.wsConn.Close()
+		c.connected.Store(false)
```

Comment:
But if the connection is not established, you should not call wsconn.Close()

....
The logic keep changing patch after patch ....

There are 2 states : connected and running :
**connected** can turned false if the sendWSMessage() failed ... and close the connection (wsConn)
**running** should only set false by Disconnect() and all loop need to terminate smoothly (without reporting error)

the chan c.quit is used as an event to break the main waiting loop (that sendMsg) and call OnDisconnected()

What is still wrong :
   o close(c.quit) need to be called just after c.running.Store(false)  
   o the if running line 146 make no sense anymore, as close(c.quit) will send a broadcast and the loop return directly
   o line 149 (before for) a c.wg.Add(1) and defer c.wd.Done() would be great but not mandatory